### PR TITLE
Fix/authentication flow

### PR DESCRIPTION
**Pull request by tanguyantoine**
## 🐻 What your PR is doing?

- improve push flow and all commands using dev portal (like integrations:create)
- fix never ending login/setup command (due to persisten connection)

## 📦 Package concerned
- @bearer/cli

## 🖥 Screenshots or screen recording

[asciicast](https://asciinema.org/a/vdwbWPmeANv0BOpWzsxMngaew)

## ✅ Checklist

- [ ] Tests were added (if necessary)
- [x] I used conventional commits

### CODE REVIEW

File: packages/cli/src/actions/createIntegration.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        const { integration } = data.data.createIntegration
+        this.logger.success('Integration successfully created')
+        this.logger.log(
+          '      name: %s\n      uuid: %s\n       Url:',
+          integration.name,
+          integration.uuid,
+          `${this.logger.constants.DeveloperPortalUrl}integrations/${integration.uuid}`
+        )
+        if (this.logger.isIntegrationLocation) {
+          // tslint:disable-next-line no-boolean-literal-compare
```
Comment (didroe) [inline packages/cli/src/actions/createIntegration.ts]: There's no literal on the next line, is this still needed?

File: packages/cli/src/actions/login.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            this.logger.log(
+              this.logger.colors.bold(`2/ when you access the success page copy the token and paste it here`)
+            )
+            const token = await askForString('Token')
+            await this.getToken(token)
+            resolve()
+          }
+        })
+      }),
+      Promise.race([
```
Comment (didroe) [inline packages/cli/src/actions/login.ts]: Do we need the nested call to `race` here?

File: packages/cli/src/actions/login.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    await Promise.race([
+      new Promise((resolve, reject) => {
+        spawned.on('close', async (code: any, signal: any) => {
+          if (code !== 0) {
+            this.logger.warn(
+              this.logger.colors.yellow(
+                `Unable to open a browser. If you want to retrieve a token please follow these steps\n`
+              )
+            )
+            this.logger.log(this.logger.colors.bold('1/ access the url below  and follow the login process:\n\n'), url)
```
Comment (didroe) [inline packages/cli/src/actions/login.ts]: There's a double space in "below  and"

File: packages/cli/src/utils/devPortal.ts [hunk 4]
```diff
@@ -14,24 +15,23 @@ export const devPortalClient = (command: BaseCommand) => {
       return Promise.reject(error)
     }
   )
+  async function requestFunction<DataReturned>(data: { query: string; variables?: any }) {
+    let token = await command.bearerConfig.getToken()
+    if (!token) {
+      command.debug('no token found, trying to log you in')
+      await promptToLogin(command)
+      // command.debug('login prompt done')
```
Comment (didroe) [inline packages/cli/src/utils/devPortal.ts]: Are you likely to want this again, or shall we remove it?

Comment (tanguyantoine) [reply packages/cli/src/utils/devPortal.ts]: I thought I removed it , 🤦‍♂

File: packages/cli/src/actions/createIntegration.ts (hunk 1 above)
Comment (tanguyantoine) [reply packages/cli/src/actions/createIntegration.ts]: good catch

File: packages/cli/src/actions/login.ts (hunk 3 above)
Comment (tanguyantoine) [reply packages/cli/src/actions/login.ts]: 👁

File: packages/cli/src/actions/login.ts (hunk 2 above)
Comment (tanguyantoine) [reply packages/cli/src/actions/login.ts]: nope, I can merge both

File: packages/cli/src/actions/createIntegration.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  uuid: string
+  name: string
+  latestActivity?: {
+    state: string
+  }
+}
+
+type TOptions = { name?: string; description?: string; link?: boolean }
+
+function randomName() {
```
Comment (RadekMolenda) [inline packages/cli/src/actions/createIntegration.ts]: :heart: keeping the `randomName` logic in dedicated module would make the createIntegration module more readable IMHO

File: packages/cli/src/actions/login.ts [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    ])
+    this.logger.success('Successfully logged in!! 🐻')
+    cliUx.action.stop()
+  }
+
+  on = (event: Event, callback: () => void) => {
+    this._listerners[event] = [...this._listerners[event], callback]
+  }
+
+  private stopServer = () => {
```
Comment (RadekMolenda) [inline packages/cli/src/actions/login.ts]:
```suggestion
  private stopServer = () => new Promise((resolve, _reject) => {
```

File: packages/cli/src/actions/login.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+  }
+}
+
+const loginFlow = createExport(LoginAction)
+
+export default loginFlow
+
+export async function promptToLogin(command: baseCommand) {
+  command.log(command.colors.bold('⚠️ It looks like you are not logged in'))
+  const { shoudlLogin } = await inquirer.prompt<{ shoudlLogin: boolean }>([
```
Comment (RadekMolenda) [inline packages/cli/src/actions/login.ts]:
```suggestion
  const { shoudLogin } = await inquirer.prompt<{ shoudLogin: boolean }>([
```

File: packages/cli/src/actions/login.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    {
+      message: 'Would you like to login?',
+      name: 'shoudlLogin',
```
Comment (RadekMolenda) [inline packages/cli/src/actions/login.ts]:
```suggestion
      name: 'shoudLogin',
```

File: packages/cli/src/actions/login.ts [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+      type: 'list',
+      choices: [{ name: 'Yes', value: true }, { name: 'No', value: false }]
+    }
+  ])
+  if (shoudlLogin) {
```
Comment (RadekMolenda) [inline packages/cli/src/actions/login.ts]:
```suggestion
  if (shoudLogin) {
```

File: packages/cli/src/commands/setup/auth.ts [hunk 10]
```diff
@@ -203,36 +202,34 @@ see examples`
   }
 
   private stopServer = () => {
-    this.debug('stopping server')
-    if (this._server) {
-      this._server.close(() => {
-        this.debug('server stopped')
-      })
-    }
+    return new Promise((resolve, _reject) => {
```
Comment (RadekMolenda) [inline packages/cli/src/commands/setup/auth.ts]: Do you think you could dry this code a bit by introducing introducing a server service class and using it in both `auth` and `login` class?

**Pull request merged**